--- uClibc-0.9.28/libc/inet/resolv.c	2005-08-18 06:49:41.000000000 +0800
+++ uClibc-0.9.28/libc/inet/resolv.c.new	2009-01-13 08:48:40.000000000 +0800
@@ -155,6 +155,8 @@
 
 #define MAX_RECURSE 5
 #define REPLY_TIMEOUT 10
+#define REPLY_TIMEOUT_STEP_ONE 1
+#define REPLY_TIMEOUT_STEP_TWO 2
 #define MAX_RETRIES 3
 #define MAX_SERVERS 3
 #define MAX_SEARCH 4
@@ -680,6 +682,7 @@
 	int variant = -1;
 	struct sockaddr_in sa;
 	int local_ns = -1, local_id = -1;
+	int resolve_step_one=1;
 #ifdef __UCLIBC_HAS_IPV6__
 	int v6;
 	struct sockaddr_in6 sa6;
@@ -787,7 +790,12 @@
 
 		FD_ZERO(&fds);
 		FD_SET(fd, &fds);
-		tv.tv_sec = REPLY_TIMEOUT;
+		if (resolve_step_one) {
+			tv.tv_sec = REPLY_TIMEOUT_STEP_ONE;
+		} else {
+			tv.tv_sec = REPLY_TIMEOUT_STEP_TWO;
+		}
+
 		tv.tv_usec = 0;
 		if (select(fd + 1, &fds, NULL, NULL, &tv) <= 0) {
 		    DPRINTF("Timeout\n");
@@ -912,9 +920,14 @@
 		   otherwise return with error */
 		{
 		    variant = -1;
-                    local_ns = (local_ns + 1) % nscount;
-                    if (local_ns == 0)
-                      retries++;
+			if (resolve_step_one) {
+				resolve_step_one = 0;
+			} else {
+				resolve_step_one = 1;
+                    		local_ns = (local_ns + 1) % nscount;                    	
+				if (local_ns == 0)
+					retries++;
+            		}
 
                     continue;
 		}
